ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractWitnessesTable.scala

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.dao.events
+
+import anorm.{BatchSql, NamedParameter}
+import com.daml.platform.store.Conversions._
+import com.daml.platform.store.DbType
+
+private[events] sealed abstract class ContractWitnessesTable {
+
+  protected val insert: String
+
+  protected val TableName = "participant_contract_witnesses"
+  protected val IdColumn = "contract_id"
+  protected val WitnessColumn = "contract_witness"
+
+  private def toNamedParameter(idAndParty: (ContractId, Party)) =
+    idAndParty match {
+      case (id, party) =>
+        Vector[NamedParameter](
+          IdColumn -> id,
+          WitnessColumn -> party,
+        )
+    }
+
+  private def prepareBatchInsert(witnesses: WitnessRelation[ContractId]): Option[BatchSql] =
+    batch(insert, Relation.flatten(witnesses).map(toNamedParameter).toSeq)
+
+  protected val delete = s"delete from $TableName where $IdColumn = {$IdColumn}"
+
+  private def prepareBatchDelete(ids: Seq[ContractId]): Option[BatchSql] =
+    batch(delete, ids.map(id => Vector[NamedParameter](IdColumn -> id)))
```

Comment:
Can't you use `Seq` (List) in these cases?

Reply:
Although, I can't be picky about it. I see it's used widely as the default immutable sequence, but, in some places (like here), the extra allocation overhead might matter.

Reply:
I don't think that's our bottleneck. 😉 
But it's a reasonable ask, on it.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/8019/commits/46b32ff5e60e015bed146cd86a8bfe5fc4de29b5